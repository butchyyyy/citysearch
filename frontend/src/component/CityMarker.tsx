import City from "model/City"
import React from "react"
import { Marker } from "react-google-maps"
import InfoWindow from "react-google-maps/lib/components/InfoWindow"
import { MarkerProps } from "react-google-maps/lib/components/Marker"

interface OwnProps {
  city: City
}

interface State {
  opened: boolean
}

type Props = Pick<MarkerProps, Exclude<keyof MarkerProps, "position">> & OwnProps

class CityMarker extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = { opened: false }
    this.toggleOpen = this.toggleOpen.bind(this)
  }

  toggleOpen() {
    this.setState({ opened: !this.state.opened })
  }

  render() {
    const { city, ...markerProps } = this.props
    return (
        <Marker
            position={{ lat: city.lat, lng: city.lng }}
            onClick={this.toggleOpen}
            {...markerProps}
        >
          {this.state.opened && (
              <InfoWindow onCloseClick={this.toggleOpen}>
                <div>
                  <strong>{city.name}</strong>
                  <br />
                  <small>{city.country}</small>
                  <br />
                  <small>Population: {city.population || "N/A"}</small>
                </div>
              </InfoWindow>
          )}
        </Marker>
    )
  }
}

export default CityMarker
