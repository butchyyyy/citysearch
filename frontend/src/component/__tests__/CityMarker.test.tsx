import { cityNullPopulation, cityPrague } from "__tests__/TestData"
import CityMarker from "component/CityMarker"
import { shallow } from "enzyme"
import React from "react"
import { InfoWindow } from "react-google-maps"

describe("Test CityMarker Component", () => {
  const wrapper = shallow(<CityMarker city={cityPrague} />)

  it("Test CityMarker Snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("Test Mouse Events", () => {
    // Test InfoWindow gets rendered on mouseover
    wrapper.simulate("mouseover")
    expect(wrapper.state("hovered")).toBeTruthy()
    expect(wrapper.find(InfoWindow).exists()).toBeTruthy()
    expect(wrapper).toMatchSnapshot()
    // Test with null population City
    wrapper.setProps({ city: cityNullPopulation })
    expect(wrapper).toMatchSnapshot()
    // Test Info window gets hidden on mouseout
    wrapper.simulate("mouseout")
    expect(wrapper.state("hovered")).toBeFalsy()
    expect(wrapper.find(InfoWindow).exists()).toBeFalsy()
  })
})
