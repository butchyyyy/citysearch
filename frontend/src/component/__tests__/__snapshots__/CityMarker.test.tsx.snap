// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CityMarker Component Test CityMarker Snapshot 1`] = `
<Marker
  onClick={[Function]}
  position={
    Object {
      "lat": 50.08804,
      "lng": 14.42076,
    }
  }
/>
`;

exports[`Test CityMarker Component Test Mouse Events 1`] = `
<Marker
  onClick={[Function]}
  position={
    Object {
      "lat": 50.08804,
      "lng": 14.42076,
    }
  }
>
  <InfoWindow
    onCloseClick={[Function]}
  >
    <div>
      <strong>
        Prague
      </strong>
      <br />
      <small>
        Czechia
      </small>
      <br />
      <small>
        Population: 
        1165581
      </small>
    </div>
  </InfoWindow>
</Marker>
`;

exports[`Test CityMarker Component Test Mouse Events 2`] = `
<Marker
  onClick={[Function]}
  position={
    Object {
      "lat": 0,
      "lng": 0,
    }
  }
>
  <InfoWindow
    onCloseClick={[Function]}
  >
    <div>
      <strong>
        I don't have a population
      </strong>
      <br />
      <small>
        Some Country
      </small>
      <br />
      <small>
        Population: 
        N/A
      </small>
    </div>
  </InfoWindow>
</Marker>
`;
